src/client/datascience/jupyter/kernels/helpers.ts

```diff
@@ -317,6 +353,18 @@ export function findPreferredKernel(
             let score = -1;
 
             if (spec) {
+                // Check if the kernel spec name matches the hash of the generated kernel spec name.
+                if (
+                    !notebookMetadata && // If we don't have metadata, only then should we compare against the interpreter.
+                    isKernelRegisteredByUs(spec) === 'newVersion' &&
+                    isPythonKernelSpec(spec) &&
+                    interpreter &&
+                    spec.name === getInterpreterKernelSpecName(interpreter)
```

Comment:
This was required, this ensure we get a perfect match based on the kernel name (name tha'ts genearted by us from the interpreter hash).

Reply:
Thinking about this. Is there a situation where two kernelspecs would be pointing to the same interpreter but one is a better match than the other? I don't think so, but just wanted to think about it some more. 

I know there are situations where more than one kernelspec is matched against the same interpreter but they likely come out to the same thing anyway.

Reply:
> Thinking about this. Is there a situation where two kernelspecs would be pointing to the same interpreter but one is a better match than the other? I don't think so, but just wanted to think about it some more.

No, in that case we'd have the kernel spec iformation in the notebook metadata, however in this `if condtion` i'm checking if notebook metadata has been provided or not.